### PR TITLE
[BEAM-8891] Spark uber jar job server

### PR DESCRIPTION
**Pull request by ibzib**
- The state and message streams got a little messy, as they're
treated as if they're independent, but at least in this implementation
the message stream is just a superset of the state stream (assuming that states last long enough to be picked up by both streams). To untangle
things a bit, I decided to a) allow duplicate messages and b) store
message history, in addition to state history.
- For unit testing, I added library `freezegun` to freeze time at 0,
which was way easier than doing a bunch of elaborate mocking just to get
one insignificant value lined up.

Follow-ups:
- The Spark REST API will return errors if there's a problem with the
job submission, but it doesn't return errors from the cluster itself
once the job is running. I couldn't figure out a good solution to this,
so I left it as a TODO(BEAM-8983).
- There should be an e2e test for this. BEAM-8984

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/python/apache_beam/runners/portability/spark_uber_jar_job_server.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+  # We only stage a single pipeline in the jar.
+  PIPELINE_NAME = 'pipeline'
+  PIPELINE_PATH = '/'.join(
+      [PIPELINE_FOLDER, PIPELINE_NAME, "pipeline.json"])
+  PIPELINE_OPTIONS_PATH = '/'.join(
+      [PIPELINE_FOLDER, PIPELINE_NAME, 'pipeline-options.json'])
+  ARTIFACT_MANIFEST_PATH = '/'.join(
+      [PIPELINE_FOLDER, PIPELINE_NAME, 'artifact-manifest.json'])
+  ARTIFACT_FOLDER = '/'.join([PIPELINE_FOLDER, PIPELINE_NAME, 'artifacts'])
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/spark_uber_jar_job_server.py]: This and everything below do not look Spark-specific. Should we create a base class for ths jar functionality?

File: sdks/python/apache_beam/runners/portability/spark_uber_jar_job_server_test.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    os.unlink(name)
+
+def spark_job():
+  return spark_uber_jar_job_server.SparkBeamJob(
+      'http://host:6066', '', '', '', '', '',
+      pipeline_options.SparkRunnerOptions())
+
+
+@unittest.skipIf(sys.version_info < (3, 6), "Requires Python 3.6+")
+class SparkUberJarJobServerTest(unittest.TestCase):
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/spark_uber_jar_job_server_test.py]: I think this could also be a shared test class for Flink and Spark.

Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/spark_uber_jar_job_server_test.py]: There's not actually that much shared code between the two, so I'm going to leave the tests as they are.

File: sdks/python/apache_beam/runners/portability/spark_uber_jar_job_server.py (hunk 1 above)
Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/spark_uber_jar_job_server.py]: Done.

**Comment by ibzib** [comment]
Run Python PreCommit

**Comment by ibzib** [comment]
Run Java PreCommit

File: sdks/python/apache_beam/runners/portability/spark_runner.py [hunk 3]
```diff
@@ -41,27 +43,23 @@ def run_pipeline(self, pipeline, options):
...
-    # TODO(BEAM-8139) submit a Spark jar to a cluster
+    spark_options = options.view_as(pipeline_options.SparkRunnerOptions)
+    if (spark_options.spark_submit_uber_jar
+        and not re.match(LOCAL_MASTER_PATTERN, spark_options.spark_master_url)):
+      if sys.version_info < (3, 6):
+        raise ValueError(
+            'spark_submit_uber_jar requires Python 3.6+, current version %s'
+            % sys.version)
+      return spark_uber_jar_job_server.SparkUberJarJobServer(
+          re.sub('^spark://', 'http://', spark_options.spark_master_url),
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/spark_runner.py]: The master and rest endpoints are distinct, so I should make the rest url an independent pipeline option before merging this PR.

**Pull request merged**